### PR TITLE
 doc: improve the description of devicetree overlay files

### LINKED ISSUE #38251: cmake: DTC_OVERLAY_FILE flags cancel board <board>.overlay files
**Is your enhancement proposal related to a problem? Please describe.**
The problem is that when DTC_OVERLAY_FILE flag is passed to cmake, it will **silently** drop `boards/<BOARD>.overlay` and `<BOARD>.overlay`.  This is due to following lines in `boilerplate.cmake`:

https://github.com/zephyrproject-rtos/zephyr/blob/bdc80ebe16c82fa589ce9c750e66d2f319123b8f/cmake/app/boilerplate.cmake#L499-L512

Documentation was not helpful. [Devicetree HOWTOs](https://docs.zephyrproject.org/latest/guides/dts/howtos.html#set-devicetree-overlays) describe `some ways to set` DTC_OVERLAY_FILE. No where it states they are mutually exclusive (my case covers 5. and 6. bulletpoint). 

**Describe the solution you'd like**
A clear documented way how are these files included and what is mutually exclusive, what is not. 
Good reference point for these kind of things is Nordic's documentation on [multi-image builds](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/1.6.0/nrf/ug_multi_image.html#image-specific-variables). 

**Describe alternatives you've considered**
If solution is not acceptable for whatever reason, at least a warning in logs should bring attention of developer that some files are being dropped.

**Additional context**
Relevant folder structure is below. I am using a build script which wraps west execution by handling different use-cases.  Config folder contains optional overlay/conf files which may or may not be used. Folder `boards/` contains overlay files which are project-specific, but I expect them to always be included whenever project is built for that board. 

This problem does not affect Kconfig overlays, only DTS. 

```bash
├── boards/
│  ├── actinius_icarus_ns.conf
│  ├── actinius_icarus_ns.overlay
├── config/
│  ├── external_flash_w25q128jv_actinius_icarus.overlay
├── src/
```

### PR DESCRIPTION
**Pull request by tejlmand**
Fixes: #38251

The current devicetree overlay description on how overlay files are
applied is not specific on the fact that the processing will stop at
the first match in the list.

For example if `DTC_OVERLAY_FILE=file.overlay` is specified, then no
devicetree overlays will be applied automatically.

If there is a match on `boards/<board>.overlay`, then the `app.overlay`
will not be applied.

This has now been made clearer in the description.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

------ 

Note: This PR depends on this PR #39030, so any comments to the first commit should go to #39030.

### CODE REVIEW

**Comment by tejlmand** [comment]
@marinjurjevic please take a look and provide feedback.

**Comment by marinjurjevic** [comment]
LGTM! 

I will take some time and go through all the problems I had with DTS/Kconfig and compile them. I hope I can come up with another draft solution which could improve documentation on issues I presented in the accompanying issue #38251

**Comment by tejlmand** [comment]
> I hope I can come up with another draft solution which could improve documentation on issues I presented in the accompanying issue #38251

Please do. As developer working daily on the system it's sometimes hard to know what is missing / causing confusion to new users.

So any suggestions for improvements are welcome.

**Comment by marinjurjevic** [comment]
> > I hope I can come up with another draft solution which could improve documentation on issues I presented in the accompanying issue #38251
> 
> Please do. As developer working daily on the system it's sometimes hard to know what is missing / causing confusion to new users.
> 
> So any suggestions for improvements are welcome.

I understand, I'm also feeling a bit biased already since I went through the process of learning all the little caveats and nitty gritty details. I do have some of the stuff written down and I'll go over them and try to see the picture from the new guy standpoint.

**Review by carlescufi — APPROVED** [review]
LGTM

**Comment by carlescufi** [comment]
@SebastianBoe @mbolivar-nordic a review would be appreciated for this one. Thanks!

File: doc/guides/dts/howtos.rst
Comment (SebastianBoe) [inline doc/guides/dts/howtos.rst:223]: We must be explaining a hundred places how to set a CMake cache variable.

Could we move this to https://docs.zephyrproject.org/latest/guides/build/index.html?highlight=cmake#build-system-cmake and refer to it from here?

File: doc/guides/dts/howtos.rst
Comment (SebastianBoe) [inline doc/guides/dts/howtos.rst:211]: The two sentences above are redundant with the description below. It should be removed IMHO.

It's better to say something once unambiguously than to say it in different ways with varying degrees of precision.

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
changes requested

File: doc/guides/dts/howtos.rst
Comment (tejlmand) [reply doc/guides/dts/howtos.rst:223]: > We must be explaining a hundred places how to set a CMake cache variable.
> 
> Could we move this to https://docs.zephyrproject.org/latest/guides/build/index.html?highlight=cmake#build-system-cmake and refer to it from here?

I see what you mean, but considering that if changing this into something like:
> By setting `DTC_OVERLAY_FILE="file1.overlay;file2.overlay;..."` in the CMake variable cache, see `<some-link>` for details.

compared to the current 3 lines, then I don't think we gain much.
We would have to:
- Describe the CMake variable to set, `DTC_OVERLAY_FILE`
- How the value is expected to look like `file1.overlay;file2.overlay;...`.
- Add a link.

but the reader reading about DTC overlay files and not knowing about CMake cache would now have to click the link, and then go back and continue reading about DTC overlay files.

A reader of this page is most likely new to Zephyr and maybe also CMake, so providing this info together makes it easier to read continuesly and not having to jump back anf forth.

So I'm not convinced that changing this would benefit the reader.

File: doc/guides/dts/howtos.rst
Comment (tejlmand) [reply doc/guides/dts/howtos.rst:211]: > It's better to say something once unambiguously than to say it in different ways with varying degrees of precision.

What I would like to avoid is confused readers, like here:
https://github.com/zephyrproject-rtos/zephyr/issues/38251

Especially this comment:
> Devicetree HOWTOs describe some ways to set DTC_OVERLAY_FILE. No where it states they are mutually exclusive 

Of course the description on how the files are processed should make it clear, just think this important enough to highlight in an extra sentence with specific purpose.

@marinjurjevic do you have any thoughts on this. I'm open to drop the extra sentence.

Comment (SebastianBoe) [reply doc/guides/dts/howtos.rst:211]: a
otherwise b
otherwise c

is now clearly communicating it is mutual exclusive IMHO.

File: doc/guides/dts/howtos.rst
Comment (SebastianBoe) [reply doc/guides/dts/howtos.rst:223]: I disagree.

We can cut down, and improve, on docs all over the place, not just here. And a user that does not intend to set DTC_OVERLAY_FILE does not need to skim generic CMake information.

Will leave this to your judgement as code owner for the build system.

File: doc/guides/dts/howtos.rst
Comment (tejlmand) [reply doc/guides/dts/howtos.rst:211]: > is now clearly communicating it is mutual exclusive IMHO.

I tend to agree with you, but as i'm not the target audience of this text I would prefer some feedback from a golden `new-comer` user.
And as @marinjurjevic was raising #38251 I see him as the golden users :wink:

File: doc/guides/dts/howtos.rst
Comment (tejlmand) [reply doc/guides/dts/howtos.rst:223]: > Will leave this to your judgement as code owner for the build system.

yeah, but this is not really build system but documentation.
So would be nice with a tech-writer kind of person to assist, as how to provide the best user experience.

**Comment by marinjurjevic** [comment]
I would like to take a step back and understand why I created #38251.
 The source of confusion on my part was that I thought `DTC_OVERLAY_FILE` is analogue to `OVERLAY_CONFIG`, but it's actually analogue to `CONF_FILE`. I was not aware of that back then, but it's clear now that there is no counterpart for `OVERLAY_CONFIG` for device tree. This is something @tejlmand agreed being confusing in general.

> Honestly, I agree with you, having `OVERLAY` in those two names `OVERLAY_CONFIG` and `DTC_OVERLAY_FILE` that have very different behavior can be very confusing.
_Originally posted by @tejlmand in https://github.com/zephyrproject-rtos/zephyr/issues/38251#issuecomment-932196029_

IMO, anything we do in the docs now won't solve the main problem -> that `DTC_OVERLAY_FILE`  variable can, in some cases, displace another file automatically -- which contradicts OVERLAY in its name. This can not happen with `OVERLAY_CONFIG`. Having same syntax but different semantics will keep confusing new users because when you're new and trying to learn new concepts you are using analogies everywhere, it's just the way people learn new stuff.

@tejlmand changes should help, as having `Otherwise` keyword will better document underlying CMake if-elseif syntax for setting `DTC_OVERLAY_FILE` variable.

File: doc/guides/dts/howtos.rst
Comment (marinjurjevic) [reply doc/guides/dts/howtos.rst:211]: I think having `Otherwise` is the key so there is no need for extra sentence. `Otherwise` should make it clear that listed options are mutually exclusive

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
changes requested

File: doc/guides/dts/howtos.rst
Comment (tejlmand) [reply doc/guides/dts/howtos.rst:211]: Thanks @SebastianBoe  and @marinjurjevic.
Will update accordingly.

Comment (tejlmand) [reply doc/guides/dts/howtos.rst]: Fixed.

**Comment by mbolivar-nordic** [comment]
> The source of confusion on my part was that I thought `DTC_OVERLAY_FILE` is analogue to `OVERLAY_CONFIG`, but it's actually analogue to `CONF_FILE`. I was not aware of that back then, but it's clear now that there is no counterpart for `OVERLAY_CONFIG` for device tree.

I disagree with this analysis.

If you want to draw an analogy, BOARD.dts is analogous to CONF_FILE, and DTC_OVERLAY_FILE is analogous to OVERLAY_CONFIG. There is exactly one board-level devicetree file (BOARD.dts in the board directory), and you extend it with overlays (via DTC_OVERLAY_FILE, e.g.). The application has a main configuration file (CONF_FILE), which you extend with fragments (OVERLAY_CONFIG). It's not a perfect analogy, but I think it's close enough.

File: doc/guides/dts/howtos.rst
Comment (mbolivar-nordic) [inline doc/guides/dts/howtos.rst:213]: This is inaccurate or at best misleading and is better explained in the documentation already linked to at the top of the section.

This text implies that the overlays themselves are all that are merged together somehow. The truth is that BOARD.dts and any overlays are concatenated and the results are treated as the final devicetree after running the C preprocessor. So all the DTS files are concatenated; it's not that just the overlays are merged. (Note the confusion earlier in the discussion about DTC_OVERLAY_FILE vs. OVERLAY_CONFIG as evidence that this text is misleading.)

This is described in detail here: https://docs.zephyrproject.org/latest/guides/dts/intro.html#input-files. Note the link to the page containing this section at the top of this HOWTO.

Furthermore, this section is duplicating content already written in the introduction. This is the HOWTOs page; it's meant as a list of recipes, not a place to introduce new ideas. I don't think we should be mixing the two here.

File: doc/guides/dts/howtos.rst
Comment (mbolivar-nordic) [inline doc/guides/dts/howtos.rst:210]: This is verbose without explaining too much. I agree a tech writer review would be better here. @FrancescoSer can you please help?

File: doc/guides/dts/howtos.rst
Comment (mbolivar-nordic) [inline doc/guides/dts/howtos.rst:215]: This is not good style; we shouldn't nest numbered lists with the same format.

```
1. hello
     1. this
     2. is a
     3. confusing list
 2. list
 ```

It's better to use something like:

```
1. ...
    1.1 ...
    1.2 ...
    1.3 ...
2. ...
```

or

```
1. ...
    i.  ...
    ii.  ...
    iii. ...
2. ...
```

etc

File: doc/guides/dts/howtos.rst
Comment (mbolivar-nordic) [reply doc/guides/dts/howtos.rst:220]: I agree with @SebastianBoe that we should be linking to content that says once and for all how to do things. That's why I wrote https://github.com/zephyrproject-rtos/zephyr/commit/5edb6d5def80322759222d514418b84c00b8c7ed (and why I'm now fighting to bring it back in https://github.com/zephyrproject-rtos/zephyr/pull/40233).

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

**Comment by marinjurjevic** [comment]
> If you want to draw an analogy, BOARD.dts is analogous to CONF_FILE, and DTC_OVERLAY_FILE is analogous to OVERLAY_CONFIG. There is exactly one board-level devicetree file (BOARD.dts in the board directory), and you extend it with overlays (via DTC_OVERLAY_FILE, e.g.). The application has a main configuration file (CONF_FILE), which you extend with fragments (OVERLAY_CONFIG). It's not a perfect analogy, but I think it's close enough.

Can you take a look at my [comment](https://github.com/zephyrproject-rtos/zephyr/issues/38251#issuecomment-932162661) on original issue. Also original issue title states my problem was with discarding an overlay file located under `boards/<board>.overlay` dir when using `DTC_OVERLAY_FILE` , but not discarding `boards/<board>.conf` file in the same folder when using `OVERLAY_CONFIG`.

Is that intended behavior? Can that be deduced just from reading [Devicetree HOWTOs](https://docs.zephyrproject.org/latest/guides/dts/howtos.html#set-devicetree-overlays)?

File: doc/guides/dts/howtos.rst
Comment (SebastianBoe) [inline doc/guides/dts/howtos.rst:208]: I actually meant that also this sentence is redundant (and less precise) with the below.

File: doc/guides/dts/howtos.rst
Comment (FrancescoSer) [reply doc/guides/dts/howtos.rst:210]: @mbolivar-nordic sure, I'll take a look

File: doc/guides/dts/howtos.rst
Comment (tejlmand) [reply doc/guides/dts/howtos.rst:215]: just following how it's done for config files:
https://docs.zephyrproject.org/latest/guides/build/kconfig/setting.html#the-initial-configuration

Comment (tejlmand) [reply doc/guides/dts/howtos.rst:215]: https://github.com/zephyrproject-rtos/zephyr/blob/119808744024d61be213b3a2e6e41738ebbc2111/doc/guides/build/kconfig/setting.rst#L140-L148
```
1. If ``CONF_FILE`` is set, the configuration file(s) specified in it are
   merged and used as the application configuration. ``CONF_FILE`` can be set
   in various ways:

   1. In :file:`CMakeLists.txt`, before calling ``find_package(Zephyr)``

   2. By passing ``-DCONF_FILE=<conf file(s)>``, either directly or via ``west``

   3. From the CMake variable cache
```

so judged by your comment, you propose to cleanup that too.

File: doc/guides/dts/howtos.rst
Comment (tejlmand) [reply doc/guides/dts/howtos.rst:213]: > This text implies that the overlays themselves are all that are merged together somehow. The truth is that BOARD.dts and any overlays are concatenated and the results are treated as the final devicetree after running the C preprocessor. 

True, internally they might be concatenated before running the preprocessor but the final DTS file will look merged.
That is, if I create an overlay for nrf52840dk as:
```
/ {
        leds {
                led0: led_0 {
                        gpios = < &gpio0 0xe 0x1 >;
                        label = "Green LED 0";
                };
        };
};
```
in order to change the pin that connects to the led, then the final zephyr.dts will contain:
```
...
      leds {
                compatible = "gpio-leds";
                led0: led_0 {
                        gpios = < &gpio0 0xe 0x1 >;
                        label = "Green LED 0";
                };
...
```
and not:
```
...
      leds {
                compatible = "gpio-leds";
                led0: led_0 {
                        gpios = < &gpio0 0xd 0x1 >;
                        label = "Green LED 0";
                };
...
/ {
        leds {
                led0: led_0 {
                        gpios = < &gpio0 0xe 0x1 >;
                        label = "Green LED 0";
                };
        };
};
```
and I think that is what the user cares about.

So I think merged is a better term than concatenated here, as concatenated could indicate they are just concatenated to form the final `zephyr.dts`.

Also merged is used here:
https://docs.zephyrproject.org/latest/guides/build/kconfig/setting.html#the-initial-configuration
> If CONF_FILE is set, the configuration file(s) specified in it are merged and used as the application configuration. CONF_FILE can be set in various ways:

where the files are processed in order and is updating the config setting to the value specified.
This means the last file specifying a given setting wins.

We could just as well have concatenated all conf files to a single file and passed that file to kconfig.py, and the last setting of a given value would win.
What is important is the final outcome `.config` / `zephyr.dts`, and there it could be seen as merged.


> Note the confusion earlier in the discussion about DTC_OVERLAY_FILE vs. OVERLAY_CONFIG as evidence that this text is misleading.

I don't see how a change from `merged` to `concatenate` makes it less confusing.
Conf files very much like device tree files whj.

For config files, the `<board>_defconfig` is not affected by `CONF_FILE` or `CONFIG_OVERLAY` settings.
Just as `<board>.dts` is not affected by the `DTS_OVERLAY_FILE`

I don't see how any re-wording here can remove that confusion to users.
For conf files we have:
- `<board>_defconfig` concatenated with `CONF_FILES` and `OVERLAY_CONFIG` to form `.config` from Kconfig tree.
  - Setting `CONF_FILES` will disable automatically selected `CONF_FILES` such as `prj.conf` and `boards/<board>.conf`
  - Setting `OVERLAY_CONFIG` will impact automatic selected `CONF_FILES`

  All files are still concatenated with `<board>_defconfig`
- `<board>.dts` concatenated with `DTS_OVERLAY_FILE`
  - Setting `DTS_OVERLAY_FILE` will disable automatically selected `DTS_OVERLAY_FILES` such as `app.dts` or `<board>.dts`

So the word `OVERLAY` has one behavior for conf files and another for devicetree.

> This is described in detail here: https://docs.zephyrproject.org/latest/guides/dts/intro.html#input-files. Note the link to the page containing this section at the top of this HOWTO.

And notice that link is inside a `tip` note describing how a `base` devicetree can be obtained.
Such link should probably be more generic link to inform the user of the different file types.
There is of course also the [BOARD.dts](https://docs.zephyrproject.org/latest/guides/dts/intro.html#devicetree-in-out-files) link, but it's not as clear to the reader that points to overlay files.

> This is the HOWTOs page; it's meant as a list of recipes, not a place to introduce new ideas. I don't think we should be mixing the two here.

i'm not changing the purpose of what is already here, just trying to clear up the existing text.
But I would agree most of this section probably belong better in the `intro` than the `howto`, so if you prefer I will move the text.

File: doc/guides/dts/howtos.rst
Comment (tejlmand) [reply doc/guides/dts/howtos.rst:220]: > I agree with @SebastianBoe that we should be linking to content that says once and for all how to do things. That's why I wrote [5edb6d5](https://github.com/zephyrproject-rtos/zephyr/commit/5edb6d5def80322759222d514418b84c00b8c7ed) (and why I'm now fighting to bring it back in #40233).

Thanks for the link.

I'm not against changing, but I think it then lacks a cleanup throughout our docset.

Just doing it for overlay files just makes the page standout from other places where CMake variable / environment variables are described, for example:
https://docs.zephyrproject.org/latest/guides/build/kconfig/setting.html#the-initial-configuration
https://docs.zephyrproject.org/latest/application/index.html#boards
https://docs.zephyrproject.org/latest/guides/modules.html#without-west

Comment (tejlmand) [reply doc/guides/dts/howtos.rst:220]: >  That's why I wrote [5edb6d5](https://github.com/zephyrproject-rtos/zephyr/commit/5edb6d5def80322759222d514418b84c00b8c7ed) (and why I'm now fighting to bring it back in #40233).

I just took a second look at that PR #40233.
That PR describes environment settings, and nothing about `-D<variable>=<value>` so linking to that will not address the comment from @SebastianBoe
> We must be explaining a hundred places how to set a CMake cache variable.

cause it would still be required to state how to set a CMake variable.

File: doc/guides/dts/howtos.rst
Comment (mbolivar-nordic) [reply doc/guides/dts/howtos.rst:215]: > so judged by your comment, you propose to cleanup that too.

If you have time, sure, I would +1, but I'm not asking for that. I am saying I don't want to add this confusing style here to the DT docs.

File: doc/guides/dts/howtos.rst
Comment (mbolivar-nordic) [reply doc/guides/dts/howtos.rst:213]: You're missing my point.

I am objecting to language saying "the overlays are merged".

I object because that omits that BOARD.dts is part of what is being merged.

**Comment by mbolivar-nordic** [comment]
Alternate PR: https://github.com/zephyrproject-rtos/zephyr/pull/40321

File: doc/guides/dts/howtos.rst
Comment (SebastianBoe) [inline doc/guides/dts/howtos.rst:205]: We are writing this in a way that makes the user believe the C preprocessor is used to implement the list of mutually exclusive methods.

Comment (FrancescoSer) [reply doc/guides/dts/howtos.rst]: @SebastianBoe Please check if the wording is now more accurate.

**Review by SebastianBoe — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
I guess we can close this now, @tejlmand ?

**Pull request closed without merging**
